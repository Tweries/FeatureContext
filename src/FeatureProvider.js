import PropTypes from 'prop-types';
import React, { createContext, useContext } from 'react';

export const FeatureContext = createContext();

export function useFeature() {
  return useContext(FeatureContext);
}

export function setFeatures(features) {
  if (Array.isArray(features)) {
    return {
      active: feature => features.indexOf(feature) !== -1
    };
  }
  throw new Error('the "features" argument must be an array');
}

function FeatureProvider({ children, features }) {
  const feature = setFeatures(features);
  return (
    <FeatureContext.Provider value={feature}>
      {children}
    </FeatureContext.Provider>
  );
}

FeatureProvider.propTypes = {
  children: PropTypes.node.isRequired,
  features: PropTypes.arrayOf(String).isRequired
};

export default FeatureProvider;
