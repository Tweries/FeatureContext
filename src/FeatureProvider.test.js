import { render } from '@testing-library/react';
import React from 'react';
import FeatureProvider, { setFeatures, useFeature } from './FeatureProvider';

describe('FeatureProvider', () => {
  const DANGER_V1 = 'DANGER_V1';
  const SUCCESS_V1 = 'SUCCESS_V1';
  const UNKNOWN_V1 = 'UNKNOWN_V1';

  function Message() {
    const feature = useFeature();
    if (feature.active(DANGER_V1)) {
      return <span>Oh Noes!</span>;
    }
    if (feature.active(SUCCESS_V1)) {
      return <span>YATTA</span>;
    }
    return (
      <span aria-label="man-shrugging" role="img">
        ü§∑‚Äç‚ôÇ
      </span>
    );
  }

  const scenarios = [
    {
      description: `${DANGER_V1} active`,
      features: [DANGER_V1]
    },
    {
      description: `${SUCCESS_V1} active`,
      feature: setFeatures([SUCCESS_V1]),
      features: [SUCCESS_V1]
    },
    {
      description: `${UNKNOWN_V1} active`,
      features: [UNKNOWN_V1]
    },
    {
      description: 'empty array',
      features: []
    }
  ];

  scenarios.forEach(({ description, features }) => {
    it(description, () => {
      const { container } = render(
        <FeatureProvider features={features}>
          <Message />
        </FeatureProvider>
      );

      expect(container).toMatchSnapshot();
    });
  });

  it('when "null" then it should throw', () => {
    expect(() => {
      render(
        <FeatureProvider features={null}>
          <Message />
        </FeatureProvider>
      );
    }).toThrowErrorMatchingSnapshot();
  });

  it('when "undefined" then it should throw', () => {
    expect(() => {
      render(
        <FeatureProvider>
          <Message />
        </FeatureProvider>
      );
    }).toThrowErrorMatchingSnapshot();
  });
});
