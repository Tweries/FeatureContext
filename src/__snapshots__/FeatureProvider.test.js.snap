// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureProvider DANGER_V1 active 1`] = `
<div>
  <span>
    Oh Noes!
  </span>
</div>
`;

exports[`FeatureProvider SUCCESS_V1 active 1`] = `
<div>
  <span>
    YATTA
  </span>
</div>
`;

exports[`FeatureProvider UNKNOWN_V1 active 1`] = `
<div>
  <span
    aria-label="man-shrugging"
    role="img"
  >
    ü§∑‚Äç‚ôÇ
  </span>
</div>
`;

exports[`FeatureProvider empty array 1`] = `
<div>
  <span
    aria-label="man-shrugging"
    role="img"
  >
    ü§∑‚Äç‚ôÇ
  </span>
</div>
`;

exports[`FeatureProvider when "null" then it should throw 1`] = `"the \\"features\\" argument must be an array"`;

exports[`FeatureProvider when "undefined" then it should throw 1`] = `"the \\"features\\" argument must be an array"`;
